File: tutorial3-complete/src-kmp/app/src/main/kotlin/dev/modaal/foyer/app/SplashBuilder.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.foyer.app

import dev.modaal.duet.kernel.KernelClock
import dev.modaal.duet.kernel.LiveClock
import dev.modaal.duet.kernel.Store
import dev.modaal.foyer.splash.SplashAction
import dev.modaal.foyer.splash.SplashDelegateEvent
import dev.modaal.foyer.splash.SplashEffectPayload
import dev.modaal.foyer.splash.SplashEnvironment
import dev.modaal.foyer.splash.SplashState
import dev.modaal.foyer.splash.makeSplashStore
import kotlinx.coroutines.CoroutineScope

/** The splash store as the app tree hands it to the render layer. */
typealias SplashStore = Store<SplashState, SplashAction, SplashEffectPayload>

/**
 * The live environment: the wall clock, and the host's sink for the delegate
 * event. The interface is the feature's; this class is the app's.
 */
private class LiveSplashEnvironment(
  private val onDelegate: (SplashDelegateEvent) -> Unit,
) : SplashEnvironment {
  override val clock: KernelClock = LiveClock

  override fun notifyHost(event: SplashDelegateEvent) = onDelegate(event)
}

/**
 * Builds one splash mount: the live environment and the store, on the scope
 * the host owns. The store's lifetime belongs to the host, which is why the
 * scope is a build parameter.
 */
class SplashBuilder {
  fun buildSplash(onDelegate: (SplashDelegateEvent) -> Unit, scope: CoroutineScope): SplashStore =
    makeSplashStore(environment = LiveSplashEnvironment(onDelegate), scope = scope)
}

File: tutorial3-start/src-ios/Libraries/FoyerKit/Sources/SplashShell/SplashBuilder.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import FoyerBridge
import FoyerKit
import Foundation

/// The live environment, implemented in Swift against the bridged Kotlin
/// interface. The framework exports the interface as an Objective-C protocol,
/// so an `NSObject` subclass conforms and the Kotlin effect handler calls
/// back into these members.
final class LiveSplashEnvironment: NSObject, SplashEnvironment {
  private let onDelegate: (SplashDelegateEvent) -> Void

  init(onDelegate: @escaping (SplashDelegateEvent) -> Void) {
    self.onDelegate = onDelegate
  }

  /// The wall clock, the Kotlin object. Under `runTest` the Kotlin lane
  /// swaps in virtual time; the app runs on this.
  var clock: KernelClock { LiveClock.shared }

  func notifyHost(event: SplashDelegateEvent) {
    onDelegate(event)
  }
}

// MARK: - Builder

/// What a mount owns. The shell owns the store's lifetime, and the mount's
/// bracket is `shell.activate()` / `shell.deactivate()`.
public final class SplashChild {
  public let shell: SplashViewShell

  init(shell: SplashViewShell) {
    self.shell = shell
  }
}

/// Builds one splash mount: scope, Kotlin store, Swift mirror, shell. The
/// same wiring every bridged feature repeats.
public final class SplashBuilder {
  public init() {}

  @MainActor
  public func buildSplash(
    onDelegate: @escaping (SplashDelegateEvent) -> Void
  ) -> SplashChild {
    let scope = mainImmediateStoreScope()
    let store = makeSplashStore(
      environment: LiveSplashEnvironment(onDelegate: onDelegate),
      scope: scope)
    let bridged = SplashKitStore(
      state: splashStateFlow(store: store),
      send: { store.send(action: $0) },
      teardown: {
        store.teardown()
        cancelStoreScope(scope: scope)
      })
    return SplashChild(shell: SplashViewShell(store: bridged))
  }
}
